iscservice/demo/util.c

```diff
@@ -0,0 +1,295 @@
+	if (rc != 0) {
+		m0_rpc_at_fini(&fop_isc->fi_args);
+		m0_rpc_at_fini(&fop_isc->fi_ret);
+		fprintf(stderr, "error! m0_rpc_at_recv() failed with %d\n", rc);
+		return rc;
+	}
+
+	m0_fop_init(arg_fop, &m0_fop_isc_fopt, fop_isc, m0_fop_release);
+	req->cir_rpc_sess = sess;
+
+	return rc;
+}
+
+void isc_req_replied(struct m0_rpc_item *item)
+{
+	int                    rc;
+	struct m0_fop         *fop = M0_AMB(fop, item, f_item);
+	struct m0_fop         *reply_fop;
+	struct m0_fop_isc_rep *isc_reply;
+	struct isc_req        *req = M0_AMB(req, fop, cir_fop);
+	const char *addr = m0_rpc_conn_addr(req->cir_rpc_sess->s_conn);
+
+	if (item->ri_error != 0) {
+		req->cir_rc = item->ri_error;
+		fprintf(stderr,
+			"No reply from %s: rc=%d.\n", addr, item->ri_error);
+		goto err;
+	}
+
+	reply_fop = m0_rpc_item_to_fop(req->cir_fop.f_item.ri_reply);
+	isc_reply = (struct m0_fop_isc_rep *)m0_fop_data(reply_fop);
+	rc = req->cir_rc = isc_reply->fir_rc;
+	if (rc != 0) {
+		fprintf(stderr,
+			"Got error in reply from %s: rc=%d.\n", addr, rc);
+		if (rc == -ENOENT)
+			fprintf(stderr, "Was isc .so library is loaded?\n");
+		goto err;
+	}
+
+	rc = m0_rpc_at_rep_get(&req->cir_isc_fop.fi_ret, &isc_reply->fir_ret,
+			       &req->cir_result);
+	if (rc != 0)
+		fprintf(stderr,
+			"rpc_at_rep_get() from %s failed: rc=%d\n", addr, rc);
+ err:
+	m0_fop_put(&req->cir_fop);
+	m0_semaphore_up(&isc_sem);
+}
+
+static const struct m0_rpc_item_ops isc_item_ops = {
+	.rio_replied = isc_req_replied,
+};
+
+static void ireqs_list_add_in_order(struct isc_req *req)
+{
+	struct isc_req *r;
+	struct m0_layout_io_plop *pl1;
+	struct m0_layout_io_plop *pl2 = M0_AMB(pl2, req->cir_plop, iop_base);
+
+	m0_list_for_each_entry(&isc_reqs, r, struct isc_req, cir_link) {
+		pl1 = M0_AMB(pl1, r->cir_plop, iop_base);
+		if (pl1->iop_goff > pl2->iop_goff)
+			break;
+	}
+	m0_list_add_before(&r->cir_link, &req->cir_link);
+}
+
+int isc_req_send(struct isc_req *req)
+{
+	int                    rc;
+	struct m0_rpc_item    *item;
+
+	item              = &req->cir_fop.f_item;
+	item->ri_session  = req->cir_rpc_sess;
+	item->ri_prio     = M0_RPC_ITEM_PRIO_MID;
+	item->ri_deadline = M0_TIME_IMMEDIATELY;
+	item->ri_nr_sent_max = M0_RPCLIB_MAX_RETRIES;
+	item->ri_ops      = &isc_item_ops;
+
+	m0_fop_get(&req->cir_fop);
+	rc = m0_rpc_post(item);
+	if (rc != 0) {
+		fprintf(stderr, "Failed to send request to %s: rc=%d\n",
+			m0_rpc_conn_addr(req->cir_rpc_sess->s_conn), rc);
+		m0_fop_put(&req->cir_fop);
+	}
+
+	ireqs_list_add_in_order(req);
+
+	return rc;
+}
+
+static void fop_fini_lock(struct m0_fop *fop)
+{
+	m0_rpc_machine_lock(m0_fop_rpc_machine(fop));
+	m0_fop_fini(fop);
+	m0_rpc_machine_unlock(m0_fop_rpc_machine(fop));
+}
+
+void isc_req_fini(struct isc_req *req)
+{
+	struct m0_fop *reply_fop = NULL;
+
+	if (req->cir_fop.f_item.ri_reply != NULL)
+		reply_fop = m0_rpc_item_to_fop(req->cir_fop.f_item.ri_reply);
+	if (reply_fop != NULL)
+		m0_fop_put_lock(reply_fop);
```

Comment:
Why do we call m0_fop_put_lock()? why not m0_fop_put()?

Reply:
Because there is M0_PRE() check at `m0_fop_put()` that rpc machine is locked.